README.md

[![Travis build
status](https://travis-ci.com/muschellij2/upstrap.svg?branch=master)](https://travis-ci.com/muschellij2/upstrap)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/muschellij2/upstrap?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/upstrap)
[![Coverage
status](https://codecov.io/gh/muschellij2/upstrap/branch/master/graph/badge.svg)](https://codecov.io/gh/muschellij2/upstrap)
<!-- README.md is generated from README.Rmd. Please edit that file -->

# upstrap

Materials accompanying the upstrap paper including vignette (.Rmd, .pdf,
.html), Sleep Heart Health Study (SHHS) data, and figure used for the
vignette

An R package called upstrap and additional code will be added as it
becomes available

# upstrap Package:

The goal of `upstrap` is to provide the upstrap resamples data with a
sample size larger than the original sample size. This method allows
users to see behavior of estimators in increasing sample sizes, commonly
done for sample size calculation.

## Installation

You can install `upstrap` from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("muschellij2/upstrap")
```


NEWS.md
# upstrap 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/upstrap.R
#' Upstrap Resampling
#'
#' @param data The data as a matrix or data frame.
#' If it is a matrix or data frame then each row
#' is considered as one multivariate observation.
#' @param statistic A function which when applied to data returns a vector
#' containing the statistic(s) of interest.
#' @param R The number of upstrap replicates for each point on the grid of
#' range multipliers. Usually this will be a single positive integer.
#' @param n.grid.multiplier number of grid points for the
#' multiplier of the sample size.
#' @param range_multiplier minimum and maximum multiplier for the sample size
#' @param ... additional arguments to pass to \code{statistic}.
#'
#' @return An object
#' @export
#'
#' @examples
#' fname = system.file("extdata", "shhs1.txt",
#' package = "upstrap")
#' # read in the data
#' data = read.table(file = fname,
#'                  header = TRUE, na.strings="NA")
#' data$MtS_SA = data$rdi4p>=15
#' statistic = function(data) {
#' fit<-glm(MtS_SA ~ gender + age_s1 + bmi_s1 +
#' HTNDerv_s1 + age_s1*HTNDerv_s1, family="binomial", data=data)
#' # could use broom
#' #  tid = broom::tidy(fit)
#' #  pval = tid$p.value[ tid$term == "HTNDerv_s1"]
#' smod = coef(summary(fit))
#' pval = smod["HTNDerv_s1", "Pr(>|z|)"]
#' pval < 0.05
#' }
#' res = upstrap(data,
#' statistic = statistic,
#' R = 50,
#' n.grid.multiplier = 1,
#' range_multiplier = c(1, 2))
#' res = colMeans(res)
upstrap = function(data, statistic, R = 10000, n.grid.multiplier = 21,
                   range_multiplier = c(1, 5), ...) {

  if (!is.function(statistic)) {
    stop("statistic needs to be a function")
  }

  # lapply(seq_len(R), function(x))
  # upstrap = statistic(data, ...)

  # set the seed for reproducibility
  # set.seed(08132018)

  # sample size of the original data
  n.oss = nrow(data)

  if (length(range_multiplier) == 1) {
    range_multiplier = rep(range_multiplier, 2)
  }

  if (length(range_multiplier) != 2) {
    warning("Range multiplier can only have 2 values, limiting to first 2")
    range_multiplier = range_multiplier[1:2]
  }
  range_multiplier = sort(range_multiplier)
  # minimum and maximum multiplier for the sample size
  # here they are set to 1 (same sample size) and 5 (5 times the original sample size)
  min.multiplier = range_multiplier[1]
  max.multiplier = range_multiplier[2]

  # set the grid of multipliers for the original sample size
  # here 1.2 stands for a sampel size that is 20% larger than the original sample size
  multiplier.grid = seq(
    from = min.multiplier,
    to = max.multiplier,
    length = max(n.grid.multiplier,
                 diff(range_multiplier) + 1)
  )

  # set the number of upstraps for each grid point;
  # this impacts running time and variability of the sample size calculation:
  # more data points, more running time, less variability
  # 10000 can take about 7-8 hours on a typical laptop
  # decrease this number for faster results
  n.upstrap = R

  # build the matrix that will contain whether or not the null hypothesis that the
  # HTN is zero for a given multiplier (r) and upstrap sample (u)
  J = length(multiplier.grid)
  check <- matrix(
    nrow = n.upstrap,
    ncol = J)

  colnames(check) = round(n.oss*multiplier.grid)
  # here j is the index of the multiplier of the original sample size, n.oss
  for (j in 1:J)
  {#Each loop corresponds to an increase in the sample size
    N =  round(n.oss*multiplier.grid[j])
    temp_index_mat <- matrix(sample.int(
      n.oss, size = N * n.upstrap,
      replace = TRUE), nrow = N, ncol = n.upstrap)

    # here u is the u-th upstrap for the r-th sample size multiplier
    # this simulation can/will be done more efficiently using matrices
    for (u in 1:n.upstrap)
    {# each loop corresponds to an upstrap with a given sample size

      # construct the upstrap sample index
      # from 1, ..., n.oss (original sample size)
      # size equal to original sample size times the multiplier
      # sampling is with replacement
      temp_index = temp_index_mat[, u]

      # extract the data (covariates and outcome) using the upstrap sample index
      temp_data <- data[temp_index,]


      # obtain the p-value for HTN in the upstrapped data
      check[u,j] <- statistic(temp_data, ...)
    }
  }
  return(check)
}


upstrap_example.R
# upstrap example: sample size calculation for ...

fname = system.file("extdata", "shhs1.txt",
                    package = "upstrap")
# read in the data
dat = read.table(file = fname,
                 header = TRUE, na.strings="NA")

# show the head of the list `dat`
head(dat)

# show the dimension of the list
dim(dat)

# to explore the data, do a scatterplot of the Body Mass Index (BMI) versus Respiratory Disturbance Index (RDI),
# variables labeled `bmi_s1` and `rdi4p`, respectively.
# The figure below provides the scatter plot of BMI versus RDI, though, for presentation purposes,
# we only show RDI less than 50.
plot(dat$bmi_s1,dat$rdi4p,pch=".",col=rgb(0,0,1,.2),cex=3,ylim=c(0,50),bty="l",
     cex.axis=1.2,col.axis="blue",main=NULL,xlab="Body Mass Index",
     ylab="Respiratory Disturbance Index (4%)")
points(mean(dat$bmi_s1,na.rm=TRUE),mean(dat$rdi4p,na.rm=TRUE),cex=2,pch=19,col=rgb(1,0,0,0.5))

# calculate the mean BMI and RDI
round(mean(dat$bmi_s1,na.rm=TRUE),digits=2)
round(mean(dat$rdi4p,na.rm=TRUE),digits=2)

## Fit a basic regression model
# define the moderate to severe sleep apnea variable from rdi4p
MtS_SA=dat$rdi4p>=15

# identify how many individuals in SHHS (visit 1) have moderate
# to severe sleep apnea
n_positive =sum(MtS_SA)
n_positive

# conduct a binary GLM regression with outcome moderate to severe
# sleep apnea (Yes=1, No=0) on a few covariates (main effects)
# and an interaction (age by HTN)
fit<-glm(MtS_SA~gender+age_s1+bmi_s1+HTNDerv_s1+age_s1*HTNDerv_s1,
         family="binomial",data=dat)
summary(fit)

# answer the question:
# For a given value of beta (say, 0.1 or 0.2), what is the sample size that
# will ensure that the null hypothesis of zero main effect of HTN will be rejected
# with a frequency at least equal to 100(1-beta)% (power)?

## The upstrap for sample size calculation

# set the seed for reproducibility
set.seed(08132018)

# sample size of the original data
n.oss=dim(dat)[1]

# number of grid points for the multiplier of the sample size
n.grid.multiplier=21

# minimum and maximum multiplier for the sample size
# here they are set to 1 (same sample size) and 5 (5 times the original sample size)
min.multiplier=1
max.multiplier=5

# set the grid of multipliers for the original sample size
# here 1.2 stands for a sampel size that is 20% larger than the original sample size
multiplier.grid=seq(from=min.multiplier, to=max.multiplier,length=n.grid.multiplier)

# set the number of upstraps for each grid point;
# this impacts running time and variability of the sample size calculation:
# more data points, more running time, less variability
# 10000 can take about 7-8 hours on a typical laptop
# decrease this number for faster results
n.upstrap=10000

# build the matrix that will contain whether or not the null hypothesis that the
# HTN is zero for a given multiplier (r) and upstrap sample (u)
check<-matrix(rep(NA,n.upstrap*n.grid.multiplier),ncol=n.grid.multiplier)

# here j is the index of the multiplier of the original sample size, n.oss
for (j in 1:n.grid.multiplier)
{#Each loop corresponds to an increase in the sample size

  # here u is the u-th upstrap for the r-th sample size multiplier
  # this simulation can/will be done more efficiently using matrices
  for (u in 1:n.upstrap)
  {# each loop corresponds to an upstrap with a given sample size

    # construct the upstrap sample index
    # from 1, ..., n.oss (original sample size)
    # size equal to original sample size times the multiplier
    # sampling is with replacement
    temp_index<-sample(1:n.oss,n.oss*multiplier.grid[j],replace=TRUE)

    # extract the data (covariates and outcome) using the upstrap sample index
    temp_data<-dat[temp_index,]
    MtS_SA=temp_data$rdi4p>=15

    # fit the same model on the upstrapped data
    fit<-glm(MtS_SA~gender+age_s1+bmi_s1+HTNDerv_s1+age_s1*HTNDerv_s1,
             family="binomial",data=temp_data)

    # obtain the p-value for HTN in the upstrapped data
    check[u,j]<-coef(summary(fit))[,4][5]<0.05
  }
}

# power check calculates the proportion of times the null hypothesis is rejected
power_check<-colMeans(check)

# plot the vector of multipliers of the original sample size, versus the power curve for
# rejecting the null hypothesis that the parameter of HTN is equal to zero
plot(multiplier.grid,power_check,type="l",col="blue",lwd=3,
     xlab="Factor by which the sample size is multiplied",
     ylab="Power to detect the HTN effect",
     bty="l",cex.axis=1.5,cex.lab=1.4,col.lab="blue",
     col.axis="blue",main=NULL)
# add horizontal lines to indicate power equal to 0.8 (orange) and 0.9 (red).
abline(h=0.8,lwd=3,col="orange")
abline(h=0.9,lwd=3,col="red")

